File: packages/core/src/app/devtools.ts
import { z } from 'zod'
import { call } from '../ipc/invoke'

/**
 * Toggle the native web inspector for the calling window — open it, or close it
 * if it is already open.
 *
 * The Rust shell compiles the inspector into every build (debug and release; see
 * `apps/desktop/src-tauri/src/devtools.rs`), so a shipped app stays debuggable.
 * Hosts without a native shell (plain-browser dev) have no inspector to toggle;
 * callers gate on `hasBridge()` before invoking.
 */
export function toggleDevtools(): Promise<void> {
  return call('toggle_devtools', {}, z.null()).then(() => undefined)
}

File: apps/desktop/src-tauri/src/devtools.rs
//! Web-inspector access — debuggable in every build, not just `cargo dev`.
//!
//! Tauri compiles the inspector in automatically for debug builds but strips it
//! from release builds unless the `tauri` crate's `devtools` feature is on (set
//! in `Cargo.toml`). With it on, the methods below exist in release too, so a
//! shipped app can always be opened up and debugged.
//!
//! The frontend binds ⌘⇧I to {@link toggle_devtools} (`@reflect/core`'s
//! `toggleDevtools`); on macOS the webview's native ⌘⌥I and right-click
//! "Inspect Element" work as well once the feature is compiled in.

use tauri::WebviewWindow;

/// Open the calling window's web inspector, or close it if it is already open.
///
/// A no-op-shaped command: it returns unit (serialized as `null`) so the typed
/// binding can validate the response like any other IPC call.
#[tauri::command]
pub fn toggle_devtools(window: WebviewWindow) {
    if window.is_devtools_open() {
        window.close_devtools();
    } else {
        window.open_devtools();
    }
}
